client/pool.go

```diff
@@ -512,3 +579,34 @@ func (pool *Pool) Close() {
 	pool.synchro.idleConnections = nil
 	pool.synchro.Unlock()
 }
+
+// checkConnection tries to connect and ping DB server
+func (pool *Pool) checkConnection(ctx context.Context) error {
+	errChan := make(chan error, 1)
+
+	go func() {
```

Comment:
I guess we need an extra goroutine because `pool.connect` does not use `pool`'s context. If we modify `pool.connect` to

```
connect: func() (*Conn, error) {
  ctx, cancel := context.WithTimeout(pool.ctx, ...)
  defer cancel()
  return ConnectWithDialer(ctx, ...)
}
```

we don't need this extra goroutine. Also it's OK to not change.

Reply:
Unfortunately, for a test case like this
```go
l, _ := net.Listen("tcp4", "127.0.0.1:0")
laddr, _ := l.Addr().(*net.TCPAddr)
//_ = l.Close() // do NOT close!

_, err = NewPoolWithOptions(laddr.String(), *testUser, *testPassword, "",
	WithNewPoolPingTimeout(time.Second),
)
```

this dialling hangs forever:
```go
ctx, cancel := context.WithTimeout(context.Background(), time.Second)
dialer := &net.Dialer{}
conn, err := ConnectWithDialer(ctx, "", addr, user, password, dbName, dialer.DialContext, po.connOptions...)
cancel()
```
This behaviour needs to be investigated separately.